api/envoy/api/v2/route/route_components.proto

```diff
@@ -1125,13 +1125,18 @@ message HedgePolicy {
   // [#not-implemented-hide:]
   type.FractionalPercent additional_request_chance = 2;
 
-  // Indicates that a hedged request should be sent when the per-try timeout
-  // is hit. This will only occur if the retry policy also indicates that a
-  // timed out request should be retried.
-  // Once a timed out request is retried due to per try timeout, the router
-  // filter will ensure that it is not retried again even if the returned
-  // response headers would otherwise be retried according the specified
-  // :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>`.
+  // Indicates that a hedged request should be sent when the per-try timeout is hit.
+  // This means that a retry will be issued without resetting the original request, leaving multiple upstream requests in flight.
+  // The first request to complete successfully will be the one returned to the caller.
+  //
+  // * At any time, a successful response (i.e. not triggering any of the retry-on conditions) would be returned to the client.
+  // * Before per-try timeout, an error response (per retry-on conditions) would be retried immediately or returned ot the client
+  //   if there are no more retries available.
+  // * After per-try timeout, an error response would be discard, as a retry in the form of a hedged request is already in progress.
+  //
+  // Note: For this to have effect, the :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>` must be one that retries on timeout
+  // (e.g. `gateway-error`).
```

Comment:
i don't think this is true, you don't need to have `gateway-error` to have it be retried. A hedged per try timeout is always retried

Reply:
@mpuncel 

I tested it and found it to be true. (In fact, wasted quite some time trying to understand why hedging doesn't work).

I actually wrote this above in reply to a similar question you raised:

Looking at this code, it seems that `retries_remaining_` would only get initialized to a non-zero value if `retry_on_` is set:
https://github.com/envoyproxy/envoy/blob/2709b6bfc73d089d823c5cde6c883806ec8f3d8e/source/common/router/retry_state_impl.cc#L121-L127

Thus `shouldRetry` would end up returning `RetryStatus::NoRetryLimitExceeded` for the hedged requests.

I can imagine this wasn't intended, since it surprised me too, but on the other hand it's congruent with how non-hedged retrying behaves.

Reply:
interesting! I do think that isn't intentional. In that case I'm not really sure how to word the comment, maybe you could say "you must have a RetryPolicy that retries at least one error code and specify the max number of retries". You don't have to have `gateway-error` specifically though it looks like from that code snippet